packages/langium/src/utils/promise-util.ts

```diff
@@ -73,7 +73,119 @@ export async function interruptAndCheck(token: CancellationToken): Promise<void>
         lastTick = current;
         await delayNextTick();
     }
-    if (token. isCancellationRequested) {
+    if (token.isCancellationRequested) {
         throw OperationCancelled;
     }
 }
+
+/**
+ * Utility class to allow for mutually exclusive read and write access.
+ *
+ * Can perform multiple read accesses at the same time, but only one write access.
+ */
+export class ReadWriteMutex {
+
+    private readCount = 0;
+    private writeCount = 0;
+
+    private writeLock: Deferred = new Deferred().resolve();
+    private readLock: Deferred = new Deferred().resolve();
+
+    private previousWriteAction?: Promise<void>;
+
+    /**
+     * Performs a single async write action, like initializing the workspace or processing document changes.
+     * Only one action will be executed at a time.
+     *
+     * If a write action is queued up while a read action is being performed, the write action will wait until all read actions have completed.
+     */
+    async write(action: () => Promise<void>): Promise<void> {
+        // Lock reading first, so that consequent read requests don't have priority over this write request
+        this.lockRead();
+        // Then await that every read previous action has completed
+        await this.readLock.promise;
+        // Append the new action to the previous action. We usually don't have to wait for long, as the previous write action
+        // 1. has either completed
+        // 2. has been cancelled due to the new write request
+        this.previousWriteAction = (this.previousWriteAction ?? Promise.resolve()).then(() =>
+            action().catch(err => {
+                if (!isOperationCancelled(err)) {
+                    console.error('Error: ', err);
+                }
+            }).finally(() => {
+                this.unlockRead();
+            })
+        );
+    }
+
+    /**
+     * Call this to lock this mutex before a write operation.
+     */
+    private lockRead(): void {
+        if (this.writeCount === 0) {
+            this.writeLock.resolve();
+            this.writeLock = new Deferred();
+        }
+        this.writeCount++;
+    }
+
+    /**
+     * Call this to unlock this mutex after a write operation.
+     */
+    private unlockRead(): void {
+        this.writeCount--;
+        if (this.writeCount === 0) {
+            this.writeLock.resolve();
+        }
+    }
+
+    /**
+     * Performs a read action on the language server. Multiple read actions can be performed at a time.
+     *
+     * If a read action is queued up while a write action is being performed, the read action will wait until the write action has completed.
+     */
+    async read<T>(action: () => Promise<T>): Promise<T> {
+        // Wait one tick for any write request to arrive
+        await delayNextTick();
+        // Await that writing has completed
+        await this.writeLock.promise;
```

Comment:
If repeated write requests were sent one after the other, couldn't this lead to the read requests never being handled? The idea of blocking all subsequent reads after a write request makes sense, but at some point it would be nice to come back and handle them. I think it may be helpful to go with a queue design instead of a priority queue. Sequenced reads can be still be performed in parallel, but writes block only the next sequence of operations, rather than blocking only subsequent reads.

Reply:
Writing happens (at least for now) only when user is typing. With auto save I could be quite frequent. With each new write on the Document the old state becomes outdated and therefor obsolet. I  think the idea is to prevent  e.g. an outline view or a validator from rendering or checking the old state. I also don't like the idea of prio ordering that much, at least not by default. I could imaging having a special request read type that is marked as low prio.

Reply:
@dhuebner Saving is unrelated to this, since the file has to be kept in sync with the editor (on every keystroke), not the file system.

Reply:
@msujew 
I see, I thought the discussion is around the Document access :)